rclrs/src/parameter/value.rs

```diff
@@ -26,27 +28,257 @@ pub enum ParameterValue {
+    ByteArray(Arc<[u8]>),
     /// An array of booleans.
     ///
     /// YAML example: `[true, false, false]`.
-    BoolArray(Vec<bool>),
+    BoolArray(Arc<[bool]>),
     /// An array of i64.
     ///
     /// YAML example: `[3, 4]`.
-    IntegerArray(Vec<i64>),
+    IntegerArray(Arc<[i64]>),
     /// An array of f64.
     ///
     /// YAML example: `[5.0, 6e2]`.
-    DoubleArray(Vec<f64>),
+    DoubleArray(Arc<[f64]>),
     /// An array of strings.
     ///
     /// YAML example: `["abc", ""]`.
-    StringArray(Vec<String>),
+    StringArray(Arc<[Arc<str>]>),
+}
+
+/// Describes the parameter's type. Similar to `ParameterValue` but also includes a `Dynamic`
+/// variant for dynamic parameters.
+#[derive(Clone, Debug, PartialEq)]
+pub enum ParameterKind {
+    /// A boolean parameter.
+    Bool,
+    /// An i64 value.
+    Integer,
+    /// An f64 value.
+    Double,
+    /// A string.
+    String,
+    /// An array of u8.
+    ByteArray,
+    /// An array of booleans.
+    BoolArray,
+    /// An array of i64.
+    IntegerArray,
+    /// An array of f64.
+    DoubleArray,
+    /// An array of strings.
+    StringArray,
+    /// A dynamic parameter that can change its type at runtime.
+    Dynamic,
+}
+
+impl From<bool> for ParameterValue {
+    fn from(value: bool) -> ParameterValue {
+        ParameterValue::Bool(value)
+    }
+}
+
+impl From<i64> for ParameterValue {
+    fn from(value: i64) -> ParameterValue {
+        ParameterValue::Integer(value)
+    }
+}
+
+impl From<f64> for ParameterValue {
+    fn from(value: f64) -> ParameterValue {
+        ParameterValue::Double(value)
+    }
+}
+
+impl From<Arc<str>> for ParameterValue {
+    fn from(value: Arc<str>) -> ParameterValue {
+        ParameterValue::String(value)
+    }
+}
+
+impl From<Arc<[u8]>> for ParameterValue {
+    fn from(value: Arc<[u8]>) -> ParameterValue {
+        ParameterValue::ByteArray(value)
+    }
+}
+
+impl From<Arc<[bool]>> for ParameterValue {
+    fn from(value: Arc<[bool]>) -> ParameterValue {
+        ParameterValue::BoolArray(value)
+    }
+}
+
+impl From<Arc<[i64]>> for ParameterValue {
+    fn from(value: Arc<[i64]>) -> ParameterValue {
+        ParameterValue::IntegerArray(value)
+    }
+}
+
+impl From<Arc<[f64]>> for ParameterValue {
+    fn from(value: Arc<[f64]>) -> ParameterValue {
+        ParameterValue::DoubleArray(value)
+    }
+}
+
+impl From<Arc<[Arc<str>]>> for ParameterValue {
+    fn from(value: Arc<[Arc<str>]>) -> ParameterValue {
+        ParameterValue::StringArray(value)
+    }
+}
+
+/// A trait that describes a value that can be converted into a parameter.
+pub trait ParameterVariant: Into<ParameterValue> + Clone {
+    /// The type used to describe the range of this parameter.
+    type Range: Into<ParameterRanges> + Default + Clone;
+    /// Attempts to convert `value` into the requested type.
+    /// Returns `Some(Self)` if the conversion was successful, `None` otherwise.
+    // TODO(luca) should we use try_from?
```

Comment:
Yes, we should

Reply:
I gave it a try but I struggled a fair bit with the error type and it ended up being a lot more verbose without really adding much (I could only get it to work by converting the errors to options through `ok()`) https://github.com/ros2-rust/ros2_rust/pull/332/commits/25cd2a8d680361c60aa9e3b62ce02b1803519aef. The main part I struggled with is the fact that the trait `TryFrom` can have any type as its `Error`, so all the code that uses `try_from(_)` needs to be able to deal with this unless we constrain the type through a quite verbose bound.
Happy to refactor / revert it, or add the trait bounds, I find them a bit tough to get right.